src/Surface.cpp

```diff
@@ -919,9 +921,9 @@ int ZCylinder::intersection(Point* point, double angle, Point* points) {
       ycurr = y0 + m * (points[num].getX() - x0);
       zcurr = z0;
       points[num].setCoords(xcurr, ycurr, zcurr);
-      if (angle < M_PI && ycurr > y0)
+      if (right && xcurr > x0)
```

Comment:
Perhaps we need some comments explaining each of these conditionals since it took you some time to dust them off and re-learn them to debug this problem in the first place.